angel_system/eval/ptg_evaluate.py

```diff
@@ -91,55 +91,115 @@ def run_eval(args):
-            det_confs_in_w.append(class_overlap['conf'].max())
-
-        # Mark detection as correct
-        for ii, r in gt_overlap.iterrows():
-            correct_label = r['class'].strip().rstrip('.')
-            correct_class_idx = labels.index(correct_label)
-
-            gt_tf_sample[correct_class_idx] = True # tp
-        
-        dets_per_valid_time_w.append(det_confs_in_w)
-        gt_true_mask.append(gt_tf_sample)
-
-    dets_per_valid_time_w = np.array(dets_per_valid_time_w)
-    gt_true_mask = np.array(gt_true_mask)
+    time_windows = np.array(time_windows)
+    dets_per_valid_time_w = np.zeros((len(time_windows), len(labels)),
+                                     dtype=float)
+
+
+    def get_time_wind_range(start, end):
+        """Return slice indices of time windows that reside completely in
+        start->end.
+
+        time_windows[ind1:ind2] all live inside start->end.
+
+        """
+        # The start time of the ith window is min_start_time + dt*i.
+        ind1_ = (start - min_start_time)/dt
+        ind1 = int(np.ceil(ind1_))
+        if ind1_ - ind1 + 1 < 1e-15:
+            # We want to avoid the case where ind1_ is (j + eps) and it gets
+            # rounded up to j + 1.
+            ind1 -= 1
+
+        # The end time of the ith window is min_start_time + dt*(i + 1).
+        ind2_ = (end - min_start_time)/dt
+        ind2 = int(np.floor(ind2_))
+        if -ind2_ + ind2 + 1 < 1e-15:
+            # We want to avoid the case where ind1_ is (j - eps) and it gets
+            # rounded up to j - 1.
+            ind1 += 1
+
+        ind1 = max([ind1, 0])
+        ind2 = min([ind2, len(time_windows)])
+
+        return ind1, ind2
+
+
+    # Valid time windows overlap with a detection.
+    valid = np.zeros(len(time_windows), dtype=bool)
+    for i in range(len(detections)):
+        ind1, ind2 = get_time_wind_range(detections['start'][i],
+                                         detections['end'][i])
+
+        valid[ind1:ind2] = True
+        correct_label = detections['class'][i].strip().rstrip('.')
+        correct_class_idx = labels.index(correct_label)
+        dets_per_valid_time_w[ind1:ind2, correct_class_idx] = np.maximum(dets_per_valid_time_w[ind1:ind2, correct_class_idx],
+                                                              detections['conf'][i])
+
+    gt_true_mask = np.zeros((len(time_windows), len(labels)), dtype=bool)
+    for i in range(len(gt)):
+        ind1, ind2 = get_time_wind_range(gt['start'][i], gt['end'][i])
+        correct_label = gt['class'][i].strip().rstrip('.')
+        correct_class_idx = labels.index(correct_label)
+
+        #print(gt['start'][i], gt['end'][i], ind1, ind2, correct_class_idx)
+
+        if np.any(gt_true_mask[ind1:ind2, :]):
+            raise Exception()
+
+        gt_true_mask[ind1:ind2, correct_class_idx] = True
+
+    if not np.all(np.sum(gt_true_mask, axis=1) <= 1):
+        raise Exception('Conflicting ground truth for same time windows')
+
+
```

Comment:
Add check that if a row isn't `gt_true_mask` isn't populated, populate 'background' label to True.